components/ads-client/src/instrument.rs

```diff
@@ -0,0 +1,179 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/.
+*/
+
+use std::sync::LazyLock;
+
+use crate::error::{ComponentError, EmitTelemetryError};
+use parking_lot::RwLock;
+use serde::{Deserialize, Serialize};
+use url::Url;
+use viaduct::Request;
+
+static DEFAULT_TELEMETRY_ENDPOINT: &str = "https://ads.mozilla.org/v1/log";
+static TELEMETRY_ENDPONT: LazyLock<RwLock<String>> =
+    LazyLock::new(|| RwLock::new(DEFAULT_TELEMETRY_ENDPOINT.to_string()));
+
+#[cfg(test)]
+pub fn set_telemetry_endpoint(endpoint: String) {
+    let mut telemetry_endpoint_lock = TELEMETRY_ENDPONT.write();
+    *telemetry_endpoint_lock = endpoint;
+}
+
+fn get_telemetry_endpoint() -> String {
+    TELEMETRY_ENDPONT.read().clone()
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum TelemetryEvent {
+    Init,
+    RenderError,
+    AdLoadError,
+    FetchError,
+    InvalidUrlError,
+}
+
+#[allow(dead_code)]
```

Comment:
we should try and avoid allowing dead code if we can - how is this used? I can't see this used anywhere in this PR. Maybe related to the mocks?